unison-cli/src/Unison/CommandLine/Completion.hs

```diff
@@ -339,3 +354,101 @@ fixupCompletion q cs@(h : t) =
+          Nothing -> pure (userHandle, Path.empty, "")
+        NamespaceListing {namespaceListingChildren} <- MaybeT $ fetchShareNamespaceInfo authHTTPClient userHandle path
+        namespaceListingChildren
+          & fmap
+            ( \case
+                Server.Subnamespace nn ->
+                  let name = Server.namespaceName nn
+                   in (NamespaceCompletion, name)
+                Server.TermObject nt ->
+                  let name = HQ'.toText $ Server.termName nt
+                   in (NamespaceCompletion, name)
+                Server.TypeObject nt ->
+                  let name = HQ'.toText $ Server.typeName nt
+                   in (TermCompletion, name)
+                Server.PatchObject np ->
+                  let name = Server.patchName np
+                   in (NamespaceCompletion, name)
+            )
+          & filter (\(typ, name) -> typ `NESet.member` completionTypes && pathSuffix `Text.isPrefixOf` name)
+          & fmap
+            ( \(_, name) ->
+                let queryPath = userHandle : coerce (Path.toList path)
+                    result = Text.unpack $ Text.intercalate "." (queryPath <> [name])
+                 in prettyCompletionWithQueryPrefix False str result
+            )
+          & pure
+
+fetchShareNamespaceInfo :: MonadIO m => AuthenticatedHttpClient -> Text -> Path.Path -> m (Maybe NamespaceListing)
+fetchShareNamespaceInfo (AuthenticatedHttpClient httpManager) userHandle path = runMaybeT do
+  let uri =
+        (Share.codeserverToURI Codeserver.defaultCodeserver)
+          { URI.uriPath = Text.unpack $ "/codebases/" <> userHandle <> "/browse",
+            URI.uriQuery =
+              if not . null $ Path.toList path
+                then Text.unpack $ "?relativeTo=" <> tShow path
+                else ""
+          }
+  req <- MaybeT $ pure (HTTP.requestFromURI uri)
+  fullResp <- liftIO $ UnliftIO.tryAny $ HTTP.httpLbs req httpManager
+  resp <- either (const empty) pure $ fullResp
+  MaybeT . pure . Aeson.decode @Server.NamespaceListing $ HTTP.responseBody resp
+
+searchUsers :: MonadIO m => AuthenticatedHttpClient -> Text -> m [Text]
+searchUsers _ "" = pure []
+searchUsers (AuthenticatedHttpClient httpManager) userHandlePrefix =
+  fromMaybe [] <$> runMaybeT do
+    let uri =
+          (Share.codeserverToURI Codeserver.defaultCodeserver)
+            { URI.uriPath = "/search",
+              URI.uriQuery = Text.unpack $ "?query=" <> userHandlePrefix
+            }
+    req <- MaybeT $ pure (HTTP.requestFromURI uri)
+    fullResp <- liftIO $ UnliftIO.tryAny $ HTTP.httpLbs req httpManager
+    resp <- either (const empty) pure $ fullResp
+    results <- (MaybeT . pure . Aeson.decode @[SearchResult] $ HTTP.responseBody resp)
+    pure $
+      results
+        & filter (\SearchResult {tag} -> tag == "User")
+        & fmap handle
+
+data SearchResult = SearchResult
```

Comment:
Here's a data-type I copied over from Enlil, should we just move it to the `unison-share-api` package instead so that Share and unison-cli are in sync?